### DISCUSSION TITLE
Handling user bindings with clj-kondo

### DISCUSSION THREAD

**Discussion by mode89** [borkdude/sci#912] [Q&A] [answered]
Hi :)

Is it possible to handle user defined macros (passed as `:bindings` to `eval-string`) with `clj-kondo` hooks?

**Comment by borkdude** [comment]
I don't understand your question, please be more elaborate or even better, give a concrete example.

**Reply by mode89** [reply]
Sorry :)

I'm working on an DSL that uses SCI as the interpreter. I have an SCI macro (let's call it `my-macro`) that I created in clojure code. I pass it into `sci.core/eval-string` through `:bindings` option. I also have a `example.clj` file which is going to be interpreted by SCI and it calls this macro. Also, I don't want to create any new namespaces and `example.clj` doesn't call `ns`, so `user` namespace is the only namespace we have.

How to make clj-kondo to understand this macro when working with `example.clj` in an editor?

I tried adding a hook (just a stub) `{:hooks {:analyze-call {my-macro my-interpreter/my-macro-hook}}}`, but clj-kondo only complains: `Unresolved symbol: my-macro`.

`:lint-as` wouldn't work for me, because the macro has a special signature.

**Reply by borkdude** [reply]
I see. You could use  `user/my-macro` perhaps?

**Reply by mode89** [reply]
When calling macro from `example.clj` by `user/my-macro`, clj-kondo says `Unresolved var: user/my-macro`.

And nothing changed after updating the hook like this `{:hooks {:analyze-call {user/my-macro my-interpreter/my-macro-hook}}}`.

**Reply by mode89** [reply]
By the way, version of clj-kondo that I use is `v2023.04.14`.

**Reply by mode89** [reply]
And sorry, I wasn't sure whether this discussion should take place on clj-kondo page or SCI page :)

**Reply by mode89** [reply]
Also tried `{:lint-as {user/my-macro clojure.core/let}}` (don't mind `let` - only for experiment), `{:lint-as {clojure.core/my-macro clojure.core/let}}`, nothing changed - clj-kondo continues complaining `Unresolved symbol: ...`

**Reply by borkdude** [reply]
Can you please make a Github repo I can clone to test your problem?

**Reply by mode89** [reply]
Sure, I will make one tomorrow.

**Reply by mode89** [reply]
[Here](https://github.com/mode89/temp-sci-and-clj-kondo/tree/master) is the repo.
Please let me know, if you need anything else.
Thank you for helping with this :)

**Reply by borkdude** [reply, marked as answer]
@mode89 Ah yes, I think clj-kondo doesn't play well with functions that aren't qualified via an ns form.

for clj-kondo users you could ask users to write this:

```
(require '[my-api :refer :all])
```

and then expose your macro/functions via `my-api` which you put in SCI's `{:namespaces {'my-api ...}}`.

When evaluating stuff in the interpreter you could automatically also evaluate `(require '[my-api :refer :all])` beforehand.

**Reply by mode89** [reply]
I see. Thank you!

Unfortunately, I can't use this approach. The DSL, that I'm making, isn't going to expose namespaces to the user. It will have Nix-like imports.

**Reply by mode89** [reply]
I don't know if it is going to fit design and goals of clj-kondo, but do you think it would be possible to have something like a pre-load script or a configuration option for declaring custom globally available bindings in clj-kondo? That would do the trick for me.

**Reply by borkdude** [reply]
I guess we could do that. Can you file a clj-kondo issue (and/or PR?)

**Reply by mode89** [reply]
Here is the issue clj-kondo/clj-kondo#2214

Not sure about making a PR by myself :)
I will definitely look into the source code, though.
